clients/src/test/java/org/apache/kafka/common/config/ConfigTransformerTest.java

```diff
@@ -0,0 +1,117 @@
+import java.util.Map;
+import java.util.Set;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ConfigTransformerTest {
+
+    private ConfigTransformer configTransformer;
+
+    @Before
+    public void setup() {
+        configTransformer = new ConfigTransformer(Collections.singletonMap("test", new TestConfigProvider()));
+    }
+
+    @Test
+    public void testReplaceVariable() throws Exception {
+        ConfigTransformerResult result = configTransformer.transform(Collections.singletonMap("myKey", "${test:testPath:testKey}"));
+        Map<String, String> data = result.data();
+        Map<String, Long> ttls = result.ttls();
+        assertEquals("testResult", data.get("myKey"));
+        assertTrue(ttls.isEmpty());
+    }
+
+    @Test
+    public void testReplaceVariableWithTTL() throws Exception {
+        ConfigTransformerResult result = configTransformer.transform(Collections.singletonMap("myKey", "${test:testPath:testKeyWithTTL}"));
+        Map<String, String> data = result.data();
+        Map<String, Long> ttls = result.ttls();
+        assertEquals("testResultWithTTL", data.get("myKey"));
+        assertEquals(1L, ttls.get("testPath").longValue());
+    }
+
+    @Test
+    public void testReplaceMultipleVariablesInValue() throws Exception {
+        ConfigTransformerResult result = configTransformer.transform(Collections.singletonMap("myKey", "hello, ${test:testPath:testKey}; goodbye, ${test:testPath:testKeyWithTTL}!!!"));
+        Map<String, String> data = result.data();
+        assertEquals("hello, testResult; goodbye, testResultWithTTL!!!", data.get("myKey"));
+    }
+
+    @Test
+    public void testNoReplacement() throws Exception {
+        ConfigTransformerResult result = configTransformer.transform(Collections.singletonMap("myKey", "${test:testPath:missingKey}"));
+        Map<String, String> data = result.data();
+        assertEquals("${test:testPath:missingKey}", data.get("myKey"));
+    }
+
+    @Test
+    public void testSingleLevelOfIndirection() throws Exception {
+        ConfigTransformerResult result = configTransformer.transform(Collections.singletonMap("myKey", "${test:testPath:testIndirection}"));
+        Map<String, String> data = result.data();
+        assertEquals("${test:testPath:testResult}", data.get("myKey"));
+    }
+
+    public static class TestConfigProvider implements ConfigProvider {
+
+        public void configure(Map<String, ?> configs) {
+        }
+
+        public ConfigData get(String path) {
+            return null;
+        }
+
+        public ConfigData get(String path, Set<String> keys) {
+            Map<String, String> data = new HashMap<>();
+            long ttl = Long.MAX_VALUE;
+            if (path.equals("testPath")) {
+                if (keys.contains("testKey")) {
+                    data.put("testKey", "testResult");
+                }
+                if (keys.contains("testKeyWithTTL")) {
```

Comment:
Might be good to use constants for these keys and values. That would be a bit more maintainable, as long as it doesn't make it unreadable.